api/circulation.py

```diff
@@ -459,6 +459,65 @@ def can_revoke_hold(self, licensepool, hold):
             return True
         return False
 
+    def _collect_event(self, patron, licensepool, name,
+                       include_neighborhood=False):
+        """Collect an analytics event.
+
+        :param patron: The Patron associated with the event. If this
+            is not specified, the current request's authenticated
+            patron will be used.
+        :param licensepool: The LicensePool associated with the event.
+        :param name: The name of the event.
+        :param include_neighborhood: If this is True, _and_ the
+            current request's authenticated patron is the same as the
+            patron in `patron`, _and_ the authenticated patron has
+            associated neighborhood information obtained from the ILS,
+            then that neighborhood information (but not the patron's
+            identity) will be associated with the circulation event.
+        """
+        if not self.analytics:
+            return
+
+        # It would be really useful to know which patron caused this
+        # this event -- this will help us get a library and
+        # potentially a neighborhood.
+        if flask.request:
+            request_patron = getattr(flask.request, 'patron', None)
+        else:
+            request_patron = None
+        patron = patron or request_patron
+
+        # We need to figure out which library is associated with
+        # this circulation event.
+        if patron:
+            # The library of the patron who caused the event.
+            library = patron.library
+        elif flask.request:
+            # The library associated with the current request.
+            library = flask.request.library
+        else:
+            # The library associated with the CirculationAPI itself.
+            library = self.library
+
+        neighborhood = None
+        if (include_neighborhood and flask.request
+            and request_patron and request_patron == patron):
```

Comment:
It's possible that no patron is passed and so patron becomes `request_patron`. In that case, `request_patron` is just being checked against itself. Is that fine?

Reply:
Yes, that's fine -- in fact that's the most common case.